examples/local/vstream_client.go

```diff
@@ -82,17 +74,375 @@ func main() {
+func storeLastVgtid(ctx context.Context, vgtid *binlogdatapb.VGtid) error {
+	_, err := json.Marshal(vgtid)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Customer is the concrete type that will be built from the stream
+type Customer struct {
+	ID    int64
+	Email string
+
+	// the fields below aren't actually in the schema, but are added for illustrative purposes
+	EmailConfirmed bool
+	Details        map[string]any
+	CreatedAt      time.Time
+}
+
+// to avoid flushing too often, we will only flush if it has been at least minFlushDuration since the last flush.
+// we're relying on heartbeat events to handle max duration between flushes, in case there are no other events.
+const minFlushDuration = 5 * time.Second
+
+// maxCustomersToFlush serves two purposes:
+//  1. it limits the number of customers that are flushed at once, to avoid large transactions. If more than this number
+//     of customers are processed, they will be flushed in chunks of this size.
+//  2. if this number is exceeded before reaching the minFlushDuration, it will trigger a flush to avoid holding too much
+//     memory in the customers slice.
+const maxCustomersToFlush = 1000
+
+func readEvents(ctx context.Context, reader vtgateconn.VStreamReader, heartbeatDur time.Duration) error {
+	// customers built from the stream will be stored in this slice, until it's time to flush them to the database,
+	// at which point they will be cleared and the slice reused
+	customers := make([]*Customer, 0, maxCustomersToFlush)
+
+	// the first events will be field events, which contains schema information for any tables that are being streamed
+	var customerFields, corderFields []*querypb.Field
+
+	// ********************************************************************************************************
+	// Heartbeat Monitoring
+	//
+	// the heartbeat ticker will be used to ensure that we haven't been disconnected from the stream. This starts
+	// a goroutine that will cancel the context if we haven't received an event in twice the heartbeat duration.
+	// We purposefully overwrite the ctx variable to ensure that we don't accidentally use the original context.
+	// ********************************************************************************************************
+	var lastEventReceivedAtUnix atomic.Int64
+
+	ctx, heartbeatCancel := context.WithCancel(ctx)
+	if heartbeatDur == 0 {
+		heartbeatDur = 30 * time.Second
+	}
+	heartbeat := time.NewTicker(heartbeatDur)
+	defer heartbeat.Stop()
+
+	go func() {
+		for {
+			select {
+			case tm := <-heartbeat.C:
+				// if we haven't received an event yet, we'll skip the heartbeat check
+				if lastEventReceivedAtUnix.Load() == 0 {
+					continue
+				}
+
+				// if we haven't received an event in twice the heartbeat duration, we'll cancel the context, since
+				// we're likely disconnected, and exit the goroutine
+				if tm.Sub(time.Unix(lastEventReceivedAtUnix.Load(), 0)) > heartbeatDur*2 {
```

Comment:
I made this heartbeat * 2 to give some buffer, but I don't have any idea if that's too much, not enough, etc